feat: improve plugin name validation error messages and field name detection (v1)

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

- Clarifies and harmonizes plugin name validation error messages across legacy and v1 plugin metadata.
- Introduces strict YAML unmarshalling for v1 plugin metadata to fail fast on unknown fields (e.g., capitalized Name).
- Coverage for valid/invalid name patterns via validPluginName regex in tests.

Why:
- Users received errors that were hard to action (e.g., “invalid name”). I think clear messages reduce confusion. We print regex expected + if the field match to nothing.
- Strict decoding for v1 catches schema mistakes early (e.g., “Name” vs “name”) while preserving backward compatibility for legacy metadata.

Backward compatibility:
- "Legacy" plugin decoding remains non-strict
- V1 manifests now fail fast on unknown fields

Related:
 - https://github.com/helm/helm/issues/31490

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
fix: improve plugin name validation error messages

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe reviewed (approved):
/lgtm

```
before> helm.sh/helm/v4/internal/plugin/metadata_legacy.go:70:Validate	62.5%
after>  helm.sh/helm/v4/internal/plugin/metadata_legacy.go:70:Validate	72.2%
```

robertsirc reviewed (dismissed):
LGTM

scottrigby commented on internal/plugin/metadata_legacy.go:
```diff
@@ -69,7 +69,10 @@ type MetadataLegacy struct {
 	if !validPluginName.MatchString(m.Name) {
-		return fmt.Errorf("invalid plugin name")
+		if m.Name == "" {
+			return fmt.Errorf("plugin name is empty or missing: ensure plugin.yaml contains 'name:' field (lowercase)")
```
`name` field is not required to be lowercase

scottrigby commented on internal/plugin/metadata_legacy_test.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func TestMetadataLegacyValidate_InvalidName(t *testing.T) {
+	metadata := MetadataLegacy{
```
here I would have a list of valid and invalid names, with a check and intended outcome for each one.

benoittgt replied:
Good idea !

benoittgt committed:
Make validation error similar and explicit for both metadatas

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix: focus only on plugin name but give more info about what we get

And improve test

Follow Scott comment:
> here I would have a list of valid and invalid names based on pattern, with a check and intended outcome for each one so that we comprehensively cover the rules.

https://github.com/helm/helm/pull/31491/files#r2524820312

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix: Make invalid name error message more similar and move tests

Follow Scott comment\
https://github.com/helm/helm/pull/31491/files#r2524933784

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix: improve plugin name validation err messages early via unmarshalling

- Add strict YAML unmarshalling for v1 plugins (d.KnownFields)
- Add comprehensive test coverage for validPluginName regex
- Maintain backwards compatibility for legacy plugins

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on internal/plugin/metadata_legacy_test.go:
```diff
@@ -0,0 +1,49 @@
+	"testing"
+)
+
+func TestMetadataLegacyValidate_PluginName(t *testing.T) {
```
Maybe this is unecessary with loader_test. I did loader_test after.

benoittgt commented:
@scottrigby thanks a lot for the very interesting feedback. I definitely went too fast toward one solution.

I added more tests to make sure we agree on the final behavior. Is enabling unmarshalling acceptable in a minor release? You mentioned it should be treated as a bug fix, so I just want to confirm.

gjenkins8 commented on internal/plugin/loader_test.go:
```diff
@@ -268,3 +268,61 @@ func TestFindPlugins(t *testing.T) {
 	}
 }
+
+func TestLoadMetadataLegacy_CapitalNameField(t *testing.T) {
```
Can we convert these into table drive tests? (stylistic, but also perhaps easier to maintain over time)

gjenkins8 commented on internal/plugin/plugin_test.go:
```diff
@@ -21,6 +21,44 @@ import (
+	tests := []struct {
+		name       string
+		pluginName string
+		shouldPass bool
```
perhaps separate into two tests?

```
testsSuccess := map[name]struct {
		pluginName string
} {
   ...
}

for testName, tt := range testsSuccess {
		t.Run(testName, func(t *testing.T) {
			assert.True(t, validPluginName.MatchString(tt.pluginName))
		})
	}

testsFail := map[name]struct {
		pluginName string
} {
   ...
}
for testName, tt := range testsFail {
		t.Run(testName, func(t *testing.T) {
            assert.False(t, validPluginName.MatchString(tt.pluginName))
		})
	}

benoittgt committed:
test: refactor TestMetadataLegacyValidate to be more generic

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
test: convert tests to table drive tests

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
test: split tests between valid and invalid

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
style: linting

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe reviewed (approved):
/lgtm

scottrigby reviewed (approved):
Nice work, lgtm :+1:
